### ISSUE TITLE
Replace PuTTY with SSH.NET

### ISSUE THREAD

**Issue by kmscode** [mRemoteNG/mRemoteNG#181]
Replace PuTTY with SSH.NET

**Comment by klonos** [comment]
Just curious, what are the major advantages of SSH.NET over PuTTY?

**Comment by jumoog** [comment]
no idea.. but maybe better Linux support?

**Comment by sparerd** [comment]
There are a few reasons we've been looking into replacing PuTTY as the default ssh/etc client:
- Tighter integration. We currently spawn a new PuTTY process off of mRemoteNG. It's essentially acting like any other external program you can define, just with a little more built-in support.
- Better security. Because PuTTY is called as an external program, we have to pass host/username/**and password** data to it in clear text over the command line.
- Better customization. Since SSH.Net is a library, we can control every minute aspect of the connection that it establishes and how it is displayed to the user.
- As @jumoog said, this also allows for better Linux support. SSH.Net is written in .Net Core which runs on Linux. Removing the PuTTY requirement is a good step toward cross platform support.

For those that still like PuTTY, we will be keeping it around as an option for handling ssh/telnet/RAW/etc. However, it will not be the default for new installations.

**Comment by sparerd** [comment]
Yea, we could still load and use putty session info from the registry. It probably wouldnt be the primary way of saving ssh connection records though

**Comment by klonos** [comment]
@sparerd thanx for taking the time to explain. All these seems pretty solid reasons as to why the change. It's just that with the change, I'm not sure whether certain features that I'm used to from PuTTY would still be available (silly things like the selection auto-copy and the right-click paste). That's all.

**Comment by jumoog** [comment]
auto-copy and the right-click paste is priceless

**Comment by igor2x** [comment]
Just my two cents. The following bug on JIRA was pointed to this Github bug:
https://mremoteng.atlassian.net/browse/MR-883

**Comment by sparerd** [comment]
This is all great feedback. We wont be replacing putty right away, there
will be plenty of time to make sure the native ssh client has the right
features

**Comment by mcarbonneaux** [comment]
must be compatible with peagent protocol for key forwarding, and then can be used with pageant or alternative like keepass with keeagent....

**Comment by kmscode** [comment]
@mcarbonneaux 

Thanks for the additional info. This is largely just a place holder at this time. We don't have any solid time frame as to when this could be implemented and due to various other factors, its unlikely that PuTTYNG will be removed (it just wouldn't be the default most likely).

**Comment by pedro2555** [comment]
@kmscode Just took a bit to read on SSH.NET, and it doesn't seem to support SSH1, which PuTTY does. Any plan on that?

**Comment by kmscode** [comment]
@pedro2555 the plan would be to keep PuTTY for things not supported by SSH.NET, but make SSH.NET the default for SSH2.

**Comment by jotatsu** [comment]
Put in 2.0 for long term usage. This would require multi clients for protocol so mRemoteNG can mix internal and external clients.

**Comment by gbakeman** [comment]
This sounds like a fun project to take on - I've made a fork and I think I'd like to work on implementing SSH.NET. I've read over this issue so far, my plans are to get SSH.NET implemented at a basic, usable level then go through and implement additional features/functionality as needed & discussed so far in this issue. Don't want to let this become a victim of feature creep.

I need to do a lot of research on SSH.NET, and familiarize myself with this project. I'll be writing up a more detailed pull request next.

**Comment by Kvarkas** [comment]
you are more than welcome to do such! :) But not to replace but to extend for initial run

**Comment by gbakeman** [comment]
Absolutely, long term I think SSH.NET would be an additional option to the traditional putty. I don't know if @jotatsu touched on this in their previous comment, but it almost seems like we need an adjustment to how different protocols are handled in mRemoteNG? I'm looking into this more, but I wonder if #242 would be considered blocking this issue.

**Comment by vallamost** [comment]
@gbakeman Were you able to make any progress on your fork to start using the Windows SSH.NET implementation over Putty? Did you get stuck on anything in particular? The modern Windows terminal's SSH is such a refreshing feeling over Putty, I would love to see it added to mRemoteNG.

**Comment by gbakeman** [comment]
I'm afraid I haven't made any progress on this. Besides other projects that've taken my time, I'm concerned there might be some larger tasks that need tackling here before we can get serious about implenenting a new connection. Apologies to anyone who might've been waiting on me 😟
